tf2onnx/rewriter/dropout_rewriter.py

```diff
@@ -60,17 +71,20 @@ def rewrite_dropout(g, ops):
                 [data],
                 outputs=[out_name],
                 name=op_name,
-                attr={"ratio": 1.0},
-                shapes=[g.get_shape(inputs2.input[0])],
-                dtypes=[g.get_dtype(inputs2.input[0])]
+                attr={"ratio": ratio},
+                shapes=[g.get_shape(input2.input[0])],
+                dtypes=[g.get_dtype(input2.input[0])]
             )
             g.replace_all_inputs(ops, outputs.output[0], new_node.output[0])
-            g.safe_remove_nodes(match.get_nodes())
-
-    # remove dropout if its ratio is 1.0
-    for node in g.get_nodes():
-        if node.type == "Dropout" and node.get_attr("ratio").f == 1.0:
-            g.replace_all_inputs(g.get_nodes(), node.output[0], node.input[0])
-            g.remove_node(node.name)
+            nodes_to_remove = []
+            for node in match.get_nodes():
+                if node.name != input3.name:
+                    nodes_to_remove.append(node)
+            if g.safe_to_remove_nodes(nodes_to_remove):
```

Comment:
we should maybe use make_sure()  and raise if this happens. Don't think tensorflow would ever give us such a graph.

Reply:
I think such a graph is technically possible if the pattern we detect is not actually from a dropout node (i.e. the user crafted the graph shape by hand from RandomUniform/floor/etc.).  Also, Keras graphs seem to sometimes output the intermediate values.

Reply:
if a tensorflow method like tf.nn.dropout() emits multiple nodes its kind of sealed, aka the nodes in the middle are not known outside.
If we mistakenly identify something that is not a dropout but handle it as dropout, no point of doing anything other than raising an exception.
Keras might put all nodes as resource types into the output but we will filter those out.

Reply:
I agree regarding Keras.  If we mistakenly match against non-dropout nodes, I believe the correct behavior is to not replace the nodes at all.  We should not create a dropout node, not raise an exception, and just directly convert all the component nodes.  I think it also makes sense to display a warning, since we expect such cases to be rare and there may be a different problem (Keras filtering may have failed).  I'll make a new commit with that behavior and see what you think.

Reply:
sure, that works as well and we do this in other rewriters.

Reply:
Also, if we can't get the ratio (if it isn't constant), we may want to similarly skip replacement (and show a warning) instead of setting it to 0.  If the user wants the ratio to be an argument, we can actually simulate that behavior by directly converting the component nodes.  What do you think?

Reply:
But than the model does not convert anymore. If you are the user - what would you prefer ?
For inference - it is ok to remove dropout regardless of rate, keras does it also this way if you run predict().
For training - sure, we should pass on the rate correctly, most likely that requires a onnx change.

Reply:
As I understand it, if we skip replacing the dropout node, the model will still successfully convert.  The constituent nodes (RandomUniform, Mul, etc.) will just be converted individually.  The result is a functioning model which perfectly matches the behavior of the original.

Reply:
we'd take a performance hit on this since now there are 2+ ops where we had 0 ops before and depending on the model we'll take that hit multiple times.
What is wrong by just removing the dropout related ops ?